server/src/main/java/com/thoughtworks/go/server/service/BuildAssignmentService.java

```diff
@@ -357,46 +356,52 @@ private Work createWork(final AgentInstance agent, final JobPlan job) {
             throw e;
         } catch (SecretResolutionFailureException e) {
             JobInstance instance = jobInstanceService.buildById(job.getJobId());
-            logSecretsResolutionFailure(job, e);
+            logSecretsResolutionFailure(job.getIdentifier(), e);
+            scheduleService.failJob(instance);
+            jobStatusTopic.post(new JobStatusMessage(job.getIdentifier(), instance.getState(), agent.getUuid()));
+            throw e;
+        } catch (RulesViolationException e) {
+            JobInstance instance = jobInstanceService.buildById(job.getJobId());
+            logRulesViolation(job.getIdentifier(), e);
             scheduleService.failJob(instance);
             jobStatusTopic.post(new JobStatusMessage(job.getIdentifier(), instance.getState(), agent.getUuid()));
-
             throw e;
         }
     }
 
-    private void logSecretsResolutionFailure(JobPlan job, SecretResolutionFailureException e) {
-        try {
-            final String description = format("\nJob for pipeline '%s' failed due to errors while resolving secret params.", job.getIdentifier().buildLocator());
-            consoleService.appendToConsoleLog(job.getIdentifier(), description);
-            consoleService.appendToConsoleLog(job.getIdentifier(), format("\nReason: %s\n", e.getMessage()));
-        } catch (IllegalArtifactLocationException | IOException e1) {
-            LOGGER.error(e1.getMessage(), e1);
+    private EnvironmentVariableContext getEnvironmentVariableContextFromEnvironment(String pipelineName) {
+        EnvironmentConfig environmentConfig = environmentConfigService.environmentForPipeline(pipelineName);
+        if (environmentConfig == null) {
+            return null;
         }
-    }
 
-    private void resolveSecretParams(BuildAssignment buildAssignment) {
-        final SecretParams allSecretParams = SecretParams.union(
-                secretParamsInMaterials(buildAssignment),
-                secretParamsInEnvironmentVariables(buildAssignment)
-        );
+        secretParamResolver.resolve(environmentConfig);
 
-        secretParamResolver.resolve(allSecretParams);
+        EnvironmentVariableContext environmentVariableContext = new EnvironmentVariableContext();
+        environmentConfig.getVariables().forEach(variable -> {
+            String resolvedValue = variable.getSecretParams().substitute(variable.getValue());
```

Comment:
Consider EnvironmentVariableConfig return a resolved value rather than the substituting here?

Reply:
Thought about it while we were implementing. The `getValue()` is used at multiple places in a different context and not all of the requires the resolved value. I can add one more method `getResolvedValue()` which returns the resolved value, and error out if has an unresolved secret param.

Reply:
Ok, or `valueForCommandline`.  guess we are using `...ForCommandline` in other places.